core/src/memory.rs

```diff
@@ -68,68 +63,83 @@ static_assertions::const_assert_eq!(WASM_PAGE_SIZE % GEAR_PAGE_SIZE, 0);
 static_assertions::const_assert_eq!(WASM_PAGE_SIZE % PAGE_STORAGE_GRANULARITY, 0);
 static_assertions::const_assert_eq!(PAGE_STORAGE_GRANULARITY % GEAR_PAGE_SIZE, 0);
 
+/// Alias for inner type of page buffer.
+pub type PageBufInner = LimitedVec<u8, (), GEAR_PAGE_SIZE>;
+
 /// Buffer for gear page data.
-#[derive(Clone, Encode, Decode, PartialEq, Eq)]
-pub struct PageBuf(Box<[u8; GEAR_PAGE_SIZE]>);
+#[derive(Clone, PartialEq, Eq, TypeInfo)]
+pub struct PageBuf(PageBufInner);
+
+// These traits are implemented intentionally by hand to achieve two goals:
+// - store PageBuf as fixed size array in a storage to eliminate extra bytes
+//      for length;
+// - work with PageBuf as with Vec. This is to workaround a limit in 2_048
+//      items for fixed length array in polkadot.js/metadata.
+//      Grep 'Only support for [Type' to get more details on that.
+impl Encode for PageBuf {
+    fn size_hint(&self) -> usize {
+        GEAR_PAGE_SIZE
+    }
+
+    fn encode_to<W: codec::Output + ?Sized>(&self, dest: &mut W) {
+        dest.write(self.0.get())
+    }
+}
+
+impl Decode for PageBuf {
+    #[inline]
+    fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+        let mut buffer = PageBufInner::filled_with(0);
+        input.read(buffer.get_mut())?;
+        Ok(Self(buffer))
+    }
+}
+
+impl codec::EncodeLike for PageBuf {}
 
 impl fmt::Debug for PageBuf {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(
             f,
             "PageBuf({:?}..{:?})",
-            &self.0[0..10],
-            &self.0[GEAR_PAGE_SIZE - 10..GEAR_PAGE_SIZE]
+            &self.0.get()[0..10],
+            &self.0.get()[GEAR_PAGE_SIZE - 10..GEAR_PAGE_SIZE]
         )
     }
 }
 
 impl Deref for PageBuf {
-    type Target = Box<[u8; GEAR_PAGE_SIZE]>;
+    type Target = [u8];
     fn deref(&self) -> &Self::Target {
-        &self.0
+        self.0.get()
     }
 }
 
 impl DerefMut for PageBuf {
     fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+        self.0.get_mut()
     }
 }
 
-impl PageBuf {
-    /// Tries to transform vec<u8> into page buffer.
-    /// Makes it without any reallocations or memcpy: vector's buffer becomes PageBuf without any changes,
-    /// except vector's buffer capacity, which is removed.
-    pub fn new_from_vec(v: Vec<u8>) -> Result<Self, Error> {
-        Box::<[u8; GEAR_PAGE_SIZE]>::try_from(v.into_boxed_slice())
-            .map_err(|data| Error::InvalidPageDataSize(data.len() as u64))
-            .map(Self)
+impl From<PageBufInner> for PageBuf {
+    fn from(mut v: PageBufInner) -> Self {
```

Comment:
I'm completely diagree with using trait `From` here. Yes, some times it's very approptiate, but not in this case. I cannot see all places where this method will be used, because of `.into()`